### PR TITLE
Llama: partial 4d masks

### LINKED ISSUE #29525: custom 4d attention masks broken by #28937
### System Info

The 4.38.2 version breaks code using custom 4d attention masks (introduced in #27539). Apparently, the custom masks gets replaced here:https://github.com/huggingface/transformers/blob/4ed9ae623d16876ad84ea89dfdf1c9378e36961b/src/transformers/models/llama/modeling_llama.py#L660-L662

The issue was introduced with #28937. It is unclear whether the relevant slow tests for 4d masks were run then, but they fail now:
```
RUN_SLOW=1 python -m pytest -v ./tests/test_modeling_utils.py::Mask4DTestFP32
FAILED tests/test_modeling_utils.py::Mask4DTestFP32::test_attention - AttributeError: 'NoneType' object has no attribute 'shape'
FAILED tests/test_modeling_utils.py::Mask4DTestFP32::test_causal_model_logits - AssertionError: Tensor-likes are not close!
FAILED tests/test_modeling_utils.py::Mask4DTestFP32::test_inner_model - AssertionError: Tensor-likes are not close!

RUN_SLOW=1 python -m pytest -v ./tests/test_modeling_utils.py::Mask4DTestFP16
FAILED tests/test_modeling_utils.py::Mask4DTestFP16::test_attention - AttributeError: 'NoneType' object has no attribute 'shape'
FAILED tests/test_modeling_utils.py::Mask4DTestFP16::test_causal_model_logits - AssertionError: Tensor-likes are not close!
```
please fix or suggest workaround

summoning @ArthurZucker 
cc @gante @younesbelkada

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

Reintroduces support for partial 4D masks in Llama (and other models with support for the static cache).

Fixes #29525 

Thank you @poedator for a clean description and a test case -- I was unaware our previous versions supported this use of 4D attention masks 🤗

### CODE REVIEW

File: tests/test_modeling_utils.py [hunk 1]
```diff
@@ -2069,3 +2071,115 @@ def test_causal_model_logits(self):
...
+                        [1, 1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0],
+                        [1, 1, 1, 0, 0, 0, 1, 1, 0, 0, 0, 0],
+                        [1, 1, 1, 0, 0, 0, 1, 1, 1, 0, 0, 0],
+                        [1, 1, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0],
+                        [1, 1, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0],
+                        [1, 1, 1, 0, 0, 0, 0, 0, 0, 1, 1, 1],
+                    ]
+                ]
+            ],
+            device="cuda:0",
```
Comment (amyeroberts) [inline tests/test_modeling_utils.py:2118]: Maybe?

```suggestion
            device=torch_device,
```

File: tests/test_modeling_utils.py [hunk 2]
```diff
@@ -2069,3 +2071,115 @@ def test_causal_model_logits(self):
...
+
+    def test_stacked_causal_mask(self):
+        # Input 0: one row per sentence; Input 1: same data, but stacked into a single row with custom attention
+        input_0, position_ids_0, input_1, mask_1, position_ids_1 = self.get_test_data()
+
+        # regular batch
+        logits_0 = self.model.forward(
+            input_0,
+            position_ids=position_ids_0,
+        ).logits
```
Comment (amyeroberts) [inline tests/test_modeling_utils.py:2136]: nit - one line

```suggestion
        logits_0 = self.model.forward(input_0, position_ids=position_ids_0).logits
```

File: tests/test_modeling_utils.py [hunk 3]
```diff
@@ -2069,3 +2071,115 @@ def test_causal_model_logits(self):
...
+        # masks
+
+        # Input 0: one row per sentence; Input 1: same data, but stacked into a single row with custom attention
+        input_0, position_ids_0, input_1, mask_1, position_ids_1 = self.get_test_data()
+
+        # regular batch
+        logits_0 = self.model.forward(
+            input_0,
+            position_ids=position_ids_0,
+        ).logits
```
Comment (amyeroberts) [inline tests/test_modeling_utils.py:2158]:
```suggestion
        logits_0 = self.model.forward(input_0, position_ids=position_ids_0).logits
```

**Review by amyeroberts — APPROVED** [review]
Thanks for fixing this and the tests ❤️

**Comment by poedator** [comment]
@gante,  thank you for the PR!
I tested it with more examples and it works OK. 

I have couple of suggestions:

- add a few lines to documentation on how to submit custom attention masks properly.

- add ability to pass a custom attention mask unchanged if its shape matches `self.causal_mask.shape`. Ideally do this before any line where causal_mask is re-created to save time and RAM. My usecase is to have a static mask for drafting model in spec decoding and only updating its elements. I'd like this to pass torch.compile.

- if you are open to extreme time saving (~ 200 µs for 4k attention_mask), the It may even be assumed that the custom mask has the right `dtype`, and has `min_dtype` in place of zeros and 0 in place of ones to avoid conversions like `(attention_mask.eq(0.0)).to(dtype=dtype) * min_dtype`

**Comment by gante** [comment]
Hi @poedator 👋 

We have other bug fixes and more impactful features in our pipeline, so I'll not work on your suggestions (at least not for now) :) However, we're always open to PRs!

**Pull request merged**